docs/root/configuration/http_conn_man/headers.rst

```diff
@@ -166,17 +165,88 @@ accurate. The source IP address of the immediate downstream node's connection to
+to XFF) are:
 
 * If *use_remote_address* is false and an XFF containing at least one IP address is
   present in the request, the trusted client address is the *last* (rightmost) IP address in XFF.
 * Otherwise, the trusted client address is the source IP address of the immediate downstream
   node's connection to Envoy.
 
+In an environment where there are one or more trusted proxies in front of an edge
+Envoy instance, the *xff_num_trusted_hops* configuration option can be used to trust
+additional addresses from XFF:
+
+* If *use_remote_address* is false and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the *N+1* th address from the right end
+  of XFF.
+* If *use_remote_address* is true and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the *N* th address from the right end
+  of XFF.
+
 Envoy uses the trusted client address contents to determine whether a request originated
 externally or internally. This influences whether the
 :ref:`config_http_conn_man_headers_x-envoy-internal` header is set.
 
+Example 1: Envoy as edge proxy, without a trusted proxy in front of it
+    Settings:
+      | use_remote_address = true
+      | xff_num_trusted_hops = 0
+
+    Request details:
+      | Downstream IP address = 192.0.2.5
+      | XFF = "198.51.100.128, 198.51.100.10, 198.51.100.1"
+
+    Result:
+      | Trusted client address = 192.0.2.5 (XFF is ignored)
+      | X-Envoy-External-Address is set to 192.0.2.5
+      | XFF is changed to "198.51.100.128, 198.51.100.10, 198.51.100.1, 192.0.2.5"
+      | Request type = external
+
+Example 2: Envoy as internal proxy, with the Envoy edge proxy from Example 1 in front of it
+    Settings:
+      | use_remote_address = false
+      | xff_num_trusted_hops = 0
+
+    Request details:
+      | Downstream IP address = 10.11.12.13 (address of the Envoy edge proxy)
+      | XFF = "198.51.100.128, 198.51.100.10, 198.51.100.1, 192.0.2.5"
+
+    Result:
+      | Trusted client address = 192.0.2.5 (last address in XFF is trusted)
+      | X-Envoy-External-Address is not modified
+      | Request type = internal
+
+
+Example 3: Envoy as edge proxy, with two trusted external proxies in front of it
+    Settings:
+      | use_remote_address = true
+      | xff_num_trusted_hops = 2
+
+    Request details:
+      | Downstream IP address = 192.0.2.5
+      | XFF = "198.51.100.128, 198.51.100.10, 198.51.100.1"
+
+    Result:
+      | Trusted client address = 198.51.100.10 (2nd to last address in XFF is trusted)
+      | X-Envoy-External-Address is set to 198.51.100.10
+      | XFF is changed to "198.51.100.128, 198.51.100.10, 198.51.100.1, 192.0.2.5"
+      | Request type = external
+
+Example 4: Envoy as internal proxy, with the edge proxy from Example 3 in front of it
+    Settings:
+      | use_remote_address = true
```

Comment:
Should this be false if we're an internal proxy?

Reply:
Yes, that's a bug on my part; I'll fix it.